### PR TITLE
[ML] Persist state for correcting loss to account for model size when training a boosted tree

### PR DESCRIPTION
**Pull request by tveasey**
We were missing state used to adjust the loss function to select the best hyperparameters based on model size. This would sometimes cause us to select different hyperparameters if the modelling failed and was restarted during training.

Note that this already tested by `testRunBoostedTreeRegressionTrainingWithStateRecovery` but only fails if the change causes us to select a different model so this doesn't reliably fail.

### CODE REVIEW

**Review by valeriy42 — APPROVED** [review]
LGTM
